core/src/main/java/org/elasticsearch/plugins/InstallPluginCommand.java

```diff
@@ -239,13 +239,35 @@ private Path unzip(Path zip, Path pluginsDir) throws IOException {
         return target;
     }
 
+    /** Checks that the plugin contain a top-level folder titled `elasticsearch` and only one sub-folder titled after the name of the plugin. */
+    private Path findPluginRoot(Path root) throws Exception {
+        Path systemRoot = root.resolve("elasticsearch");
+        if (!Files.exists(systemRoot)) {
+            throw new UserError(CliTool.ExitStatus.USAGE, "'elasticsearch' directory is missing in the plugin zip");
+        }
+        if(!Files.isDirectory(systemRoot)) {
+            throw new UserError(CliTool.ExitStatus.USAGE, "'elasticsearch' must be a directory in the plugin zip");
+        }
+        Path[] pluginsRoot = FileSystemUtils.files(systemRoot);
+        if (pluginsRoot.length != 1) {
+            throw new UserError(CliTool.ExitStatus.USAGE, "'elasticsearch' directory must contain only one sub-folder titled after the name of the plugin");
+        }
+        return pluginsRoot[0];
+    }
+
     /** Load information about the plugin, and verify it can be installed with no errors. */
     private PluginInfo verify(Path pluginRoot, Environment env) throws Exception {
         // read and validate the plugin descriptor
         PluginInfo info = PluginInfo.readFromProperties(pluginRoot);
+
+        // check if the name of the plugin matches with the directory name inside the zip
+        String dirName = pluginRoot.getFileName().toString();
+        if (!dirName.equals(info.getName())) {
+            throw new UserError(CliTool.ExitStatus.USAGE, "plugin name '" + info.getName() + "' does not match the directory name '" + dirName + "' inside the zip");
+        }
         terminal.println(VERBOSE, info.toString());
 
-        // don't let luser install plugin as a module...
+        // don't let user install plugin as a module...
```

Comment:
I thought it was a typo as well until I read https://en.wikipedia.org/wiki/Luser :p